First implementation of real-valued linspace. (#420)
fortran-lang/stdlib: doc/specs/stdlib_math.md

  ! clipped_value <- 3.02500010
end program demo_clip_real
```

### `linspace` - Create a linearly spaced rank one array

#### Description

Returns a linearly spaced rank 1 array from [`start`, `end`]. Optionally, you can specify the length of the returned array by passing `n`.

#### Syntax

`res = [[stdlib_math(module):linspace(interface)]] (start, end [, n])`

#### Status

Experimental

#### Class

> function.

Reviewer: ```suggestion
Function.
```

#### Argument(s)

`start`: Shall be scalar of any numeric type. This argument is `intent(in)`.
`end`: Shall be the same `type` and `kind` as `start`. This argument is `intent(in)`.

> `n`: Shall be an integer specifying the length of the output. This argument is `intent(in)`.

Reviewer: ```suggestion
`n`: Shall be an integer specifying the length of the output. This argument is `optional` and `intent(in)`.
```

#### Output value or Result value

The output is a rank 1 array of `type` and `kind`, whose length is either 100 (default value) or `n`.

If `n` == 1, return a rank 1 array whose only element is `end`.
If `n` <= 0, return a rank 1 array with length 0

> 

Reviewer: It should include what will happen with integer arguments.

Author: Included in latest snapshot

#### Examples

##### Example 1:

Here inputs are of type `complex` and kind `dp`
```fortran
program demo_linspace_complex
  use stdlib_math, only: linspace
  use stdlib_kinds, only: dp
  implicit none

  complex(dp) :: start = complex(10.0_dp, 5.0_dp)
  complex(dp) :: end = complex(-10.0_dp, 15.0_dp)

  complex(dp) :: z(11)

  z = linspace(start, end, 11)
end program demo_linspace_complex
```

##### Example 2:

Here inputs are of type `integer` and kind `int16`
```fortran
program demo_linspace_int16
  use stdlib_math, only: linspace
  use stdlib_kinds, only: int16
  implicit none

  integer(int16) :: start = 10_int16
  integer(int16) :: end = 23_int16

  integer(int16) :: r(15)

  r = linspace(start, end, 15)
end program demo_linspace_int16

> ```

Reviewer: I would say that one of the 2 examples is enough. Or you could try to merge both.

Author: Tbh I would like to keep the second example (albeit modified so that `r` is a `real(dp)` array) to make the behaviour when  passing integers abundantly clear. Having an extra example can't hurt.

### `logspace` - Create a logarithmically spaced rank one array

#### Description

> Returns a logarithmically spaced rank 1 array from [`base`^`start`, `base`^`end`]. Optionally, you can specify the length of the returned array by passing `n`.

Reviewer: ```suggestion
Returns a logarithmically spaced rank 1 array from [`base`^`start`, `base`^`end`]. The default size of the array is 100. Optionally, you can specify the length of the returned array by passing `n`.
```

#### Syntax

`res = [[stdlib_math(module):logspace(interface)]] (start, end [, n [, base]])`

#### Status

Experimental

#### Class

> function.

Reviewer: ```suggestion
Function.
```

#### Argument(s)

> `start`: Shall be scalar of any numeric type. This argument is `intent(in)`.

Reviewer: ```suggestion
`start`: Shall be a scalar of any numeric type. This argument is `intent(in)`.
```

`end`: Shall be the same `type` and `kind` as `start`. This argument is `intent(in)`.

> `n`: Shall be an integer specifying the length of the output. This argument is `intent(in)`.

Reviewer: ```suggestion
`n`: Shall be an integer specifying the length of the output. This argument is `optional` and `intent(in)`.
```

> `base` : Shall be a scalar of any numeric type. This argument is `intent(in)`

Reviewer: ```suggestion
`base` : Shall be a scalar of any numeric type. This argument is `optional` and `intent(in)`
```

#### Output value or Result value

The output is a rank 1 array of `type` and `kind`, whose length is either 50 (default value) or `n`.

If `n` == 1, return a rank 1 array whose only element is `base`^`end`.

> If `n` <= 0, return a rank 1 array with length 0

Reviewer: It should iunclude what will happen with integer/real/complex arguments

Author: Agreed, I'm including a simple markdown table to make that clear.

Author: ![image](https://user-images.githubusercontent.com/40283612/125348300-095aa880-e32a-11eb-8f3e-bbec2a9914af.png)
 Here is the rendered table.

#### Examples

##### Example 1:

Here inputs are of type `complex` and kind `dp`
```fortran
program demo_logspace_complex
  use stdlib_math, only: logspace
  use stdlib_kinds, only: dp
  implicit none

  complex(dp) :: start = complex(10.0_dp, 5.0_dp)
  complex(dp) :: end = complex(-10.0_dp, 15.0_dp)

  complex(dp) :: z(11)

  z = logspace(start, end, 11)
end program demo_logspace_complex
```

##### Example 2:

Here inputs are of type `integer` and kind `int16`
```fortran
program demo_logspace_int16
  use stdlib_math, only: logspace
  use stdlib_kinds, only: int16
  implicit none

  integer(int16) :: start = 10_int16
  integer(int16) :: end = 23_int16

  integer(int16) :: r(15)

  r = logspace(start, end, 15)
end program demo_logspace_int16

> ```

Reviewer: Same remark as for the previous section.

  ! clipped_value <- 3.02500010
end program demo_clip_real
```

### `linspace` - Create a linearly spaced rank one array

#### Description

Returns a linearly spaced rank 1 array from [`start`, `end`]. Optionally, you can specify the length of the returned array by passing `n`.

#### Syntax

`res = [[stdlib_math(module):linspace(interface)]] (start, end [, n])`

#### Status

Experimental

#### Class

Function.

#### Argument(s)

`start`: Shall be scalar of any numeric type or kind. This argument is `intent(in)`.
`end`: Shall be the same `type` and `kind` as `start`. This argument is `intent(in)`.
`n`: Shall be an integer specifying the length of the output. This argument is `optional` and `intent(in)`.

#### Output value or Result value

The output is a rank 1 array whose length is either 100 (default value) or `n`.

If `n` == 1, return a rank 1 array whose only element is `end`.
If `n` <= 0, return a rank 1 array with length 0.

If `start`/`end` are real or complex types, the `result` will be the same `type` and `kind` as `start`/`end`.

> If `start`/`end` are integer types, the `result` will default to a double precision real array.

Author: I'm confused. You say that `dp` might not refer to `double precision`, but then you are the one suggesting replacing `double precision` with `real(dp)`. Am I missing something here?

> If `start`/`end` are integer types, the `result` will default to a double precision real array.

Reviewer: In `stdlib_math`, the result is declared as `real(dp) :: res(n)`, with `dp` being provided by `stdlib_kinds`. `dp` is the same as `real64` provided by `iso_fortran_env` [here](https://github.com/fortran-lang/stdlib/blob/492543d0086ff3821ecf2ee8f0a9cf46a945a7ac/src/stdlib_kinds.f90#L3). However, `real64` does not garanty that it is `double precision`. See [this post by Doctor Fortran](https://stevelionel.com/drfortran/2017/03/27/doctor-fortran-in-it-takes-all-kinds/) for more details.

Therefore, I think it is more correct to use `real(dp)` in your sentence, instead of `double precision`.

> If `start`/`end` are integer types, the `result` will default to a double precision real array.

Author: Awesome! Thanks for the clarification, I didn't understand the nuance between `dp` and `douple precision` but what you've said here clears that up. Currently incorporating your corrections :)

> If `start`/`end` are integer types, the `result` will default to a double precision real array.

Author: Also, I'm sorry to keep bugging but can you approve the workflow checks? And if all of them pass, would you kindly merge this PR? I'm just so excited to be a first-time contributor!!! This is totally a tangent but as I mentioned I'll be starting a Comp Sci program in the fall, and this will be my first symbolic step in my programming career! And I can't wait!!!

  ! clipped_value <- 3.02500010
end program demo_clip_real
```

### `linspace` function

#### Description

> Returns a linearly spaced rank 1 array from [`start`, `end`]. Optionally, you can specify the length of the returned array by passing `n`.

Author: I'll incorporate this change with the tiny correction "can be specified"

> Returns a linearly spaced rank 1 array from [`start`, `end`]. Optionally, you can specify the length of the returned array by passing `n`.

Author: Actually, I'm going to push back a little bit on this change. According to google developer's [technical writing course](https://developers.google.com/tech-writing/one/active-voice), "The vast majority of sentences in technical writing should be in active voice."

> Returns a linearly spaced rank 1 array from [`start`, `end`]. Optionally, you can specify the length of the returned array by passing `n`.

Reviewer: I don't like active voice in technical writing. But it is a personal taste. So keep as it is.

> Returns a linearly spaced rank 1 array from [`start`, `end`]. Optionally, you can specify the length of the returned array by passing `n`.

Reviewer: ```suggestion
Returns a linearly spaced rank 1 array from [`start`, `end`]. The default size of the array is 100. Optionally, you can specify the length of the returned array by passing `n`.
```

#### Syntax

`res = [[stdlib_math(module):linspace(interface)]] (start, end [, n])`

#### Status

Experimental

#### Class

function.

#### Argument(s)

> `start`: scalar of any numeric type. This argument is `intent(in)`.  

Reviewer: ```suggestion
`start`: Shall be scalar of any numeric type. This argument is `intent(in)`.  
```

> `end`: same `type` and `kind` as `start`. This argument is `intent(in)`.  

Reviewer: ```suggestion
`end`: Shall be the same `type` and `kind` as `start`. This argument is `intent(in)`.  
```

> `n`: Integer parameter specifying the length of the output. This argument is `intent(in)`.

Reviewer: ```suggestion
`n`: Shall be an integer specifying the length of the output. This argument is `intent(in)`.
```
